File: python/scenes/yellow_buddy.py
#!/usr/bin/env python3
"""Yellow Buddy OBJ character scene for LumenCore."""
from __future__ import annotations

import os
import sys
from pathlib import Path

import lumencore as lc


def resolve_asset(relative: str) -> str:
    candidates = []
    root = os.environ.get("LUMENCORE_ROOT")
    if root:
        candidates.append(str(Path(root) / relative))
    candidates.extend(
        [
            relative,
            str(Path("..") / relative),
            str(Path("../..") / relative),
            str(Path("/work") / relative),
        ]
    )
    for path in candidates:
        if Path(path).is_file():
            return path
    return candidates[0]


def main() -> int:
    out = sys.argv[1] if len(sys.argv) > 1 else "yellow_buddy.png"
    spp = int(sys.argv[2]) if len(sys.argv) > 2 else 256
    denoise = (int(sys.argv[3]) != 0) if len(sys.argv) > 3 else True

    scene = lc.Scene()
    yellow = scene.add_material(lc.Material(base_color=(0.95, 0.82, 0.12), roughness=0.45))
    overalls = scene.add_material(lc.Material(base_color=(0.15, 0.35, 0.75), roughness=0.55))
    strap = scene.add_material(lc.Material(base_color=(0.08, 0.08, 0.08), roughness=0.7))
    goggle = scene.add_material(lc.Material(base_color=(0.7, 0.7, 0.75), metallic=0.85, roughness=0.15))
    lens = scene.add_material(
        lc.Material(base_color=(0.75, 0.88, 0.98), roughness=0.05, transmission=0.85, ior=1.5)
    )
    eye_white = scene.add_material(lc.Material(base_color=(0.95, 0.95, 0.95), roughness=0.4))
    pupil = scene.add_material(lc.Material(base_color=(0.05, 0.05, 0.05), roughness=0.6))
    boot = scene.add_material(lc.Material(base_color=(0.18, 0.1, 0.06), roughness=0.75))
    floor_mat = scene.add_material(lc.Material(base_color=(0.55, 0.55, 0.58), roughness=0.85))
    wall = scene.add_material(lc.Material(base_color=(0.82, 0.8, 0.78), roughness=0.9))
    chrome = scene.add_material(lc.Material(base_color=(0.95, 0.95, 0.98), metallic=1.0, roughness=0.05))

    mtl_map = {
        "Yellow": yellow,
        "Overalls": overalls,
        "Strap": strap,
        "Goggle": goggle,
        "Lens": lens,
        "EyeWhite": eye_white,
        "Pupil": pupil,
        "Boot": boot,
    }

    obj_path = resolve_asset("assets/models/yellow_buddy.obj")
    buddy = lc.load_obj(obj_path, mtl_map, yellow)
    buddy = lc.transform_mesh(buddy, (0, 0, 0), (1, 1, 1), (0, 0.35, 0))
    scene.add_mesh(buddy)
    print(f"Loaded {obj_path}")

    scene.add_mesh(lc.make_quad((-4, 0, -4), (8, 0, 0), (0, 0, 8), floor_mat))
    scene.add_mesh(lc.make_quad((-4, 0, -2.2), (8, 0, 0), (0, 4, 0), wall))
    scene.add_mesh(lc.make_uv_sphere((-1.35, 0.35, 0.8), 0.35, chrome))
    scene.add_mesh(lc.make_box((1.1, 0.0, 0.4), (1.7, 0.7, 1.0), overalls))

    light_corner = (-1.2, 3.6, -0.5)
    light_u = (2.4, 0, 0)
    light_v = (0, 0, 2.0)
    light_mat = scene.add_material(lc.Material(base_color=(0, 0, 0), roughness=1.0, emission=(18, 17, 15)))
    scene.add_mesh(lc.make_quad(light_corner, light_u, light_v, light_mat))
    scene.add_quad_light(light_corner, light_u, light_v, (18, 17, 15))

    scene.background_top = (0.55, 0.68, 0.88)
    scene.background_bottom = (0.35, 0.35, 0.4)

    camera = lc.Camera(eye=(2.6, 1.55, 3.4), lookat=(0.0, 1.0, 0.15), fov_y_deg=35, aspect=16 / 9)
    cfg = lc.RenderConfig(width=2560, height=1440, spp=spp, denoise=denoise, output_path=out)
    lc.Renderer().render(scene, camera, cfg)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/yellow_buddy.cc
#include "nrtx/nrtx.h"

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace nrtx;

static bool file_exists(const std::string &path) {
  std::ifstream in(path);
  return static_cast<bool>(in);
}

static std::string resolve_asset(const std::string &relative) {
  std::vector<std::string> candidates;
  if (const char *env = std::getenv("LUMENCORE_ROOT")) {
    candidates.push_back(std::string(env) + "/" + relative);
  }
  candidates.push_back(relative);
  candidates.push_back("../" + relative);
  candidates.push_back("../../" + relative);
  candidates.push_back("/work/" + relative);
  for (const std::string &path : candidates) {
    if (file_exists(path)) {
      return path;
    }
  }
  return candidates.front();
}

int main(int argc, char **argv) {
  const std::string out = argc > 1 ? argv[1] : "yellow_buddy.png";
  const int spp = argc > 2 ? std::atoi(argv[2]) : 256;
  const bool denoise = argc > 3 ? std::atoi(argv[3]) != 0 : true;

  Scene scene;

  const int yellow =
      scene.add_material({make_float3(0.95f, 0.82f, 0.12f), 0, 0.45f, 0, 1.5f, make_float3(0, 0, 0)});
  const int overalls =
      scene.add_material({make_float3(0.15f, 0.35f, 0.75f), 0, 0.55f, 0, 1.5f, make_float3(0, 0, 0)});
  const int strap =
      scene.add_material({make_float3(0.08f, 0.08f, 0.08f), 0, 0.7f, 0, 1.5f, make_float3(0, 0, 0)});
  const int goggle =
      scene.add_material({make_float3(0.7f, 0.7f, 0.75f), 0.85f, 0.15f, 0, 1.5f, make_float3(0, 0, 0)});
  const int lens =
      scene.add_material({make_float3(0.75f, 0.88f, 0.98f), 0, 0.05f, 0.85f, 1.5f, make_float3(0, 0, 0)});
  const int eye_white =
      scene.add_material({make_float3(0.95f, 0.95f, 0.95f), 0, 0.4f, 0, 1.5f, make_float3(0, 0, 0)});
  const int pupil =
      scene.add_material({make_float3(0.05f, 0.05f, 0.05f), 0, 0.6f, 0, 1.5f, make_float3(0, 0, 0)});
  const int boot =
      scene.add_material({make_float3(0.18f, 0.1f, 0.06f), 0, 0.75f, 0, 1.5f, make_float3(0, 0, 0)});
  const int floor_mat =
      scene.add_material({make_float3(0.55f, 0.55f, 0.58f), 0, 0.85f, 0, 1.5f, make_float3(0, 0, 0)});
  const int wall =
      scene.add_material({make_float3(0.82f, 0.8f, 0.78f), 0, 0.9f, 0, 1.5f, make_float3(0, 0, 0)});
  const int chrome =
      scene.add_material({make_float3(0.95f, 0.95f, 0.98f), 1.0f, 0.05f, 0, 1.5f, make_float3(0, 0, 0)});

  std::unordered_map<std::string, int> mtl_map = {
      {"Yellow", yellow},   {"Overalls", overalls}, {"Strap", strap}, {"Goggle", goggle},
      {"Lens", lens},       {"EyeWhite", eye_white}, {"Pupil", pupil}, {"Boot", boot},
  };

  const std::string obj_path = resolve_asset("assets/models/yellow_buddy.obj");
  try {
    Mesh buddy = load_obj(obj_path, mtl_map, yellow);
    buddy = transform_mesh(buddy, make_float3(0.0f, 0.0f, 0.0f), make_float3(1.0f, 1.0f, 1.0f),
                           make_float3(0.0f, 0.35f, 0.0f));
    scene.add_mesh(std::move(buddy));
    std::cout << "Loaded " << obj_path << "\n";
  } catch (const std::exception &ex) {
    std::cerr << "Failed to load character OBJ (" << obj_path << "): " << ex.what() << "\n";
    return 1;
  }

  scene.add_mesh(make_quad(make_float3(-4, 0, -4), make_float3(8, 0, 0), make_float3(0, 0, 8), floor_mat));
  scene.add_mesh(make_quad(make_float3(-4, 0, -2.2f), make_float3(8, 0, 0), make_float3(0, 4, 0), wall));
  scene.add_mesh(make_uv_sphere(make_float3(-1.35f, 0.35f, 0.8f), 0.35f, chrome));
  scene.add_mesh(make_box(make_float3(1.1f, 0.0f, 0.4f), make_float3(1.7f, 0.7f, 1.0f), overalls));

  const float3 light_corner = make_float3(-1.2f, 3.6f, -0.5f);
  const float3 light_u = make_float3(2.4f, 0, 0);
  const float3 light_v = make_float3(0, 0, 2.0f);
  const int light_mat =
      scene.add_material({make_float3(0, 0, 0), 0, 1, 0, 1.5f, make_float3(18.0f, 17.0f, 15.0f)});
  scene.add_mesh(make_quad(light_corner, light_u, light_v, light_mat));
  scene.add_quad_light(light_corner, light_u, light_v, make_float3(18.0f, 17.0f, 15.0f));

  scene.background_top = make_float3(0.55f, 0.68f, 0.88f);
  scene.background_bottom = make_float3(0.35f, 0.35f, 0.4f);

  Camera camera;
  camera.eye = make_float3(2.6f, 1.55f, 3.4f);
  camera.lookat = make_float3(0.0f, 1.0f, 0.15f);
  camera.fov_y_deg = 35.0f;
  camera.aspect = 16.0f / 9.0f;

  RenderConfig cfg;
  cfg.width = 1280;
  cfg.height = 720;
  cfg.spp = spp;
  cfg.denoise = denoise;
  cfg.output_path = out;

  try {
    Renderer renderer;
    renderer.render(scene, camera, cfg);
  } catch (const std::exception &ex) {
    std::cerr << "Error: " << ex.what() << "\n";
    return 1;
  }
  return 0;
}
